### PR TITLE
Appflowy integration and pins editing flow

### PR DESCRIPTION
**Pull request by gtalha07**
Refs -> #1273.
_**(Best to be reviewed commit by commit)**_
From a superseded #1288, in order to put PR on short basis and easy reviewal, the changes include:
- Integrate Appflowy rich text editor for content editing. This feature can now be accessed from Acter Labs and only supports Pins for now with limited functionality.
- Simplified Pins editing flow by removing side sheet method and instead provides option for edit mode .i.e. can be enabled from action menu of pin item. 
- Fixes related to ```ActerPin``` object updates (rust). 
- Minor typo improvements and corrections of methods. 
- e2e integration test (flutter):
   - [x] Edit Pin Example ensuring changes are reflected.
   - [x] Ensure Acter Labs pin rich text editor selection is enabled/disabled.
   
Pin Editing and Lab Editor Demo:

(Desktop) -> https://github.com/acterglobal/a3/assets/68579938/90840ef4-347d-4303-92f8-1a4e277aaf45
(Mobile) -> https://github.com/acterglobal/a3/assets/68579938/c3da777c-7087-4beb-af50-ce9a0f46c3c6

### CODE REVIEW

**Comment by gnunicorn** [comment]
@kumarpalsinh25 flutter lint complaints

File: app/lib/features/chat/models/chat_list_state/chat_list_state.freezed.dart
Comment (gnunicorn) [inline app/lib/features/chat/models/chat_list_state/chat_list_state.freezed.dart:1]: This changes seems unrelated!?!

File: app/lib/features/pins/pages/pin_page.dart [hunk 1]
```diff
@@ -30,39 +28,37 @@ class PinPage extends ConsumerWidget {
...
       if (memb.canString('CanRedactOwn') &&
           memb.userId().toString() == pin.sender().toString()) {
         final roomId = pin.roomIdStr();
-        actions.addAll([
+        actions.add(
           PopupMenuItem(
             onTap: () => showAdaptiveDialog(
               context: context,
               builder: (context) => RedactContentWidget(
                 title: 'Remove this post',
```
Comment (gnunicorn) [inline app/lib/features/pins/pages/pin_page.dart:59]:
```suggestion
                title: 'Remove this pin',
```

just noticed...

File: app/lib/features/pins/pages/pin_page.dart [hunk 2]
```diff
@@ -119,117 +115,69 @@ class PinPage extends ConsumerWidget {
...
+                  decoration: const BoxDecoration(gradient: primaryGradient),
+                ),
+                leading: IconButton(
+                  onPressed: () => context.canPop()
+                      ? context.pop()
+                      : context.goNamed(Routes.pins.name),
+                  icon: const Icon(
+                    Icons.chevron_left,
+                    size: 42,
                   ),
```
Comment (gnunicorn) [inline app/lib/features/pins/pages/pin_page.dart:150]:
```suggestion
                automaticallyImplyLeading: true,
                centerTitle: false,
                leadingWidth: 40,
                toolbarHeight: 100,
                flexibleSpace: Container(
                  decoration: const BoxDecoration(gradient: primaryGradient),
                ),
```

Don't provide a custom `leading` if it isn't necessary. And it isn't necessary. Always prefer `automaticallyImplyLeading` over adding custom local routing and designs as they are forgotten easily when updates need to happen and then lead to weird custom behaviors.

File: app/lib/features/pins/pages/pin_page.dart [hunk 3]
```diff
@@ -119,117 +115,69 @@ class PinPage extends ConsumerWidget {
...
-                          ),
-                        ),
-                        Padding(
-                          padding: const EdgeInsets.all(8),
-                          child: Column(children: content),
-                        ),
-                      ],
-                    ),
+                ),
+                title: TextFormField(
```
Comment (gnunicorn) [inline app/lib/features/pins/pages/pin_page.dart:152]: why always make that a text form field? Doesn't that mean it won't overflow for text larger than the width? Why not just ...

```suggestion
                title: !pinEdit.editMode ?Text(acterPin.title()` :  TextFormField(
```

File: app/lib/features/pins/providers/notifiers/edit_state_notifier.dart
Comment (gnunicorn) [inline app/lib/features/pins/providers/notifiers/edit_state_notifier.dart:1]: I commented before that I don't quite get why we should be using the complex, verbose and error-prone riverpod for form editing when a much simpler Form+FormField+validator is sufficient. Still waiting for a comment on that ...

File: app/lib/features/pins/providers/notifiers/pins_notifiers.dart [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    extends AutoDisposeFamilyAsyncNotifier<ActerPin, String> {
+  late Stream<void> _listener;
+
+  Future<ActerPin> _getPin() async {
+    final client = ref.watch(alwaysClientProvider);
+    try {
+      return await client.pin(arg);
+    } catch (e) {
+      return await client.waitForPin(arg, null);
+    }
```
Comment (gnunicorn) [inline app/lib/features/pins/providers/notifiers/pins_notifiers.dart:34]: going for waitForPin right away is enough. it does internally the same but more efficient.

```suggestion
      return await client.waitForPin(arg, null);
```

File: app/lib/features/pins/sheets/edit_pin_sheet.dart
Comment (gnunicorn) [inline app/lib/features/pins/sheets/edit_pin_sheet.dart:1]: :tada: bye, bye old friend!

File: app/lib/features/pins/widgets/pin_item.dart
Comment (gnunicorn) [inline app/lib/features/pins/widgets/pin_item.dart:1]: Do we have e2ee tests for this? I don't see them having been updated... do they still run?

File: native/acter/src/api/pins.rs [hunk 5]
```diff
@@ -224,8 +226,8 @@ impl Pin {
             .and_then(|t| t.formatted.clone().map(|f| f.body))
     }
 
-    pub fn content_text(&self) -> Option<String> {
-        self.content.content.clone().map(|t| t.body)
+    pub fn content(&self) -> Option<MsgContent> {
```
Comment (gnunicorn) [inline native/acter/src/api/pins.rs:229]: :+1:

File: native/acter/src/api/pins.rs [hunk 6]
```diff
@@ -361,6 +363,12 @@ impl PinDraft {
         self
     }
 
+    pub fn content_html(&mut self, body: String, html_body: String) -> &mut Self {
```
Comment (gnunicorn) [inline native/acter/src/api/pins.rs:366]: :+1:

**Review by gnunicorn — CHANGES_REQUESTED** [review]
There's still a few areas I need some feedback on please...

File: app/lib/features/pins/widgets/pin_item.dart
Comment (gtalha07) [reply app/lib/features/pins/widgets/pin_item.dart:1]: Yeah I'll add that now, previously I thought it'd more ideal if we add it after pin content refactor .i.e. adding attachments and favicons but that can be extended along the way as it'd take some thoughtful implementation but the sooner, the better.

File: app/lib/features/pins/providers/notifiers/edit_state_notifier.dart
Comment (gtalha07) [reply app/lib/features/pins/providers/notifiers/edit_state_notifier.dart:1]: I think I put forward my take on this on superseded PR , that is its easier to manage state variables when residing in one model and class, having multiple text controllers with state listening being handled by multiple providers is hard to manage or if we want some migration in future. In this case, controllers will be in different widgets .i.e the title controller is on ```PinPage``` rather than item widget. @kumarpalsinh25 What's your opinion on this?

File: app/lib/features/pins/pages/pin_page.dart (hunk 1 above)
Comment (gtalha07) [reply app/lib/features/pins/pages/pin_page.dart:59]: This has been fixed in [deebdb4](https://github.com/acterglobal/a3/pull/1366/commits/deebdb49c86fe50f9119df2060e87aaa0da9e9c5) :)

**Review by kumarpalsinh25 — CHANGES_REQUESTED** [review]
@gtalha07 
Share some suggestion I have added.
We can improve code and UI-UX part later stage.

I think we can proceed for `merge to main` after we don't have any bugs in functional behaviour.

Suggest:
- Share end-to-end manual testing screen-recording video related to newly affected pin features.

Really appreciate your patience and hard-work. 👏

File: app/lib/features/pins/pages/pin_page.dart [hunk 7]
```diff
@@ -119,117 +115,69 @@ class PinPage extends ConsumerWidget {
...
     );
   }
 
   @override
   Widget build(BuildContext context, WidgetRef ref) {
-    // ignore: unused_local_variable
     final pin = ref.watch(pinProvider(pinId));
+
     return Scaffold(
       body: CustomScrollView(
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/pins/pages/pin_page.dart:134]: @gnunicorn 
Understood your point and we definitely need to handle views with scroll whenever required.

- `SingleChildScrollView` is the best way to manage this. Use it at the component level or screen level based on the requirements. Like wrap the column or any component where scroll will be required or we can wrap entire page body with it to manage screen wide scrolling.

- `CustomScrollView` is the best option when we need to create various scrolling effects with the use of slivers.
Screenshot 2024-02-12 at 12 30 15 PM

`SingleChildScrollView` is a good choice for simple lists. But, if you need to create a more complex list, then you should use Sliver widgets.

Here are some examples of when you might use Sliver widgets:
- You want to create a grid of widgets.
- You want to create a list with headers and footers.
- You want to create a list with different scroll behaviors, such as a list that can be scrolled in multiple directions.

File: app/integration_test/tests/pins.dart [hunk 8]
```diff
@@ -12,7 +18,35 @@ import '../support/super_invites.dart';
 import '../support/util.dart';
 
 extension ActerNews on ConvenientTest {
-  Future<void> createPin(String spaceId, String text, String url) async {
+  Future<void> gotoPin(String pinId, {Key? appTab}) async {
```
Comment (kumarpalsinh25) [inline app/integration_test/tests/pins.dart:21]:
```suggestion
  Future<void> goToPin(String pinId, {Key? appTab}) async {
```

File: app/lib/features/pins/pages/pin_page.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/pins/pages/pin_page.dart:134]: Sure @gnunicorn, I will add this to style docs.

**Review by kumarpalsinh25 — APPROVED** [review]
LGTM!
